jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,432 @@
+                getServer().getThreadPool().execute(entry);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private class Entry implements CyclicTimeouts.Expirable, Runnable
+    {
+        private final Request request;
+        private final Response response;
+        private final Callback callback;
+        private final int priority;
+        private final long expireNanoTime;
+
+        private Entry(Request request, Response response, Callback callback, int priority)
+        {
+            this.request = request;
+            this.response = response;
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            long suspendNanos = NanoTime.now() + maxSuspend.toNanos();
+            if (suspendNanos == Long.MAX_VALUE)
+                --suspendNanos;
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : suspendNanos;
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return expireNanoTime;
+        }
+
+        private void expire()
+        {
+            // The request timed out, therefore it never acquired a permit.
+            boolean removed = queues.get(priority).remove(this);
+            if (removed)
+            {
+                // See correspondent state machine logic in handle() and resume().
+                state.getAndIncrement();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} timeout {}", QoSHandler.this, request);
+                request.setAttribute(EXPIRED_ATTRIBUTE_NAME, true);
+                failSuspended(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503, new TimeoutException());
+            }
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                boolean handled = forward(request, response, callback);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} handled={} {}", QoSHandler.this, handled, request);
+                if (!handled)
+                    failSuspended(request, response, callback, HttpStatus.NOT_FOUND_404, null);
+            }
+            catch (Throwable x)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} failed {}", QoSHandler.this, request, x);
+                failSuspended(request, response, callback, HttpStatus.INTERNAL_SERVER_ERROR_500, x);
+            }
+        }
+    }
+
+    private class Resumer extends HttpStream.Wrapper
+    {
+        private final Request request;
+
+        private Resumer(HttpStream wrapped, Request request)
+        {
+            super(wrapped);
+            this.request = request;
+        }
+
+        @Override
+        public void succeeded()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} succeeded {}", QoSHandler.this, request);
+            resume();
```

Comment:
I'm just thinking out-loud here for future optimizations.... If the thread here was the same thread that called `handle` and that then called `handleWithPermit`, then we could delay calling `resume()` here and somehow set state so that when we return to handle, it will then call a new `resumeWithPermit` method.

I think that might work for a very common case.  But I'm not sure how much it would gain.